Books api request

SoleCincis opened the pull request:
Current state : 
![IMG_5281E51A7387-1](https://user-images.githubusercontent.com/43641321/140789427-d6c3a063-c329-430e-9c40-da6130e33702.jpeg)

SoleCincis committed:
feat(BooksScreen): created BooksScreen

- first API request is successful ; list of books is there each with its title

SoleCincis committed:
feat(BooksScreen): created BooksScreen

- first API request is successful ; list of books is there each with its title

SoleCincis committed:
feat(BooksScreen): showing title and first author + plus number of extra authors

- condition for 0 more authors is missing
- ISBAN is missing

SoleCincis committed:
feat(BooksScreen): showing title and first author + plus number of extra authors

- condition for 0 more authors is missing
- ISBAN is missing

SoleCincis committed:
feat(BooksScreen): showing data on list

- showing first author + number of more authors if there is some
- isbn still not visible on screen
- add some style to check for better readability

SoleCincis committed:
feat(BooksScreen): showing data on list

- showing first author + number of more authors if there is some
- isbn still not visible on screen
- add some style to check for better readability

SoleCincis commented on screens/BooksScreen.js:
```diff
@@ -21,28 +21,69 @@ export default function BooksScreen() {
+  });
+
+  const isbnArray = availabilities.map((availability)=>{
+    console.log("availability.isbn:", availability !== undefined ? availability.isbn : '')
```
output: 
<img width="256" alt="Screenshot 2021-11-01 at 17 22 37" src="https://user-images.githubusercontent.com/43641321/139706016-3cacc8b1-9827-4d17-a684-cc683842f443.png">
availability.isbn are different types

SoleCincis commented on screens/BooksScreen.js:
```diff
@@ -21,28 +21,69 @@ export default function BooksScreen() {
+    console.log("availability.isbn:", availability !== undefined ? availability.isbn : '')
+    return availability !== undefined ? availability.isbn : ''
+  })
+    console.log("isbnArray:", isbnArray)
```
output: <img width="163" alt="Screenshot 2021-11-01 at 17 22 52" src="https://user-images.githubusercontent.com/43641321/139706349-d469fbf9-4fea-483f-9a13-0a8d9f479b8c.png">
isbnArray is indeed an array of isbn different values

SoleCincis commented on screens/BooksScreen.js:
```diff
@@ -21,28 +21,69 @@ export default function BooksScreen() {
+
+  const isbn = isbnArray.forEach((isbn) => {
+    console.log("ISBN:",  typeof isbn === "string" ? isbn : '')
+    return typeof isbn === "string" ? `[ISBN: ${isbn}]` : ''
```
output: <img width="132" alt="Screenshot 2021-11-01 at 17 22 21" src="https://user-images.githubusercontent.com/43641321/139706638-0ce7a718-3a24-41bd-9525-f4b8331e8811.png">
finally I can see just sting values if isbn

SoleCincis commented on screens/BooksScreen.js:
```diff
@@ -21,28 +21,69 @@ export default function BooksScreen() {
+    console.log("ISBN:",  typeof isbn === "string" ? isbn : '')
+    return typeof isbn === "string" ? `[ISBN: ${isbn}]` : ''
+  })
+  console.log("constISBN:", isbn)
```
output: <img width="200" alt="Screenshot 2021-11-01 at 17 23 07" src="https://user-images.githubusercontent.com/43641321/139706990-f42a2d34-fec3-4d29-8296-4dbea1583e52.png">

digorath replied:
[Array.map](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/map) vs [Array.foreach](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/foreach)

This should point you in the right direction to understand what is currently going on.

We discussed the difference last time we spoke.

SoleCincis replied:
finally solved in 02bf2fed85b876743338afb291853980ce0fcc3f. 
There is no need to loop through data to get isbn, meaning I'm not using neither .map or .forEach

SoleCincis commented on screens/BooksScreen.js:
```diff
@@ -21,28 +21,69 @@ export default function BooksScreen() {
+          <Text>
+            {typeof isbn === "string" ? isbn : '' }
+          </Text>
+        )} */}
```
because const isbn it's not what I need to pass to my list . 
I tried to loop trough isbn directly in my renderItem function. 
No success isbn still not visible , nothing happens.

digorath replied:
see in [this discussion](https://github.com/SoleCincis/tracker-App/pull/33/files#r740356746)

SoleCincis commented on screens/BooksScreen.js:
```diff
@@ -21,28 +21,69 @@ export default function BooksScreen() {
+      {isLoading ? <ActivityIndicator /> : (
+        <FlatList
           data={data}
           key={({ id }, index) => id}
```
this is not giving me issues or warnings . Cannot find the example where I copied it from, but in RN docu for flatlist it's also a function https://reactnative.dev/docs/flatlist

digorath replied:
I can not find a `key` property of the `FlatList` in the documentation you referred to.

SoleCincis replied:
from code example and here [keyextractor](https://reactnative.dev/docs/flatlist#keyextractor) is "explained". To be honest I understand just the first sentence there..
<img width="820" alt="Screenshot 2021-11-02 at 14 34 41" src="https://user-images.githubusercontent.com/43641321/139857116-037b4e6a-5bd0-4ce5-bca7-5caab5c74909.png">

digorath replied:
there is no `key` prop in your example.

SoleCincis replied:
I guess key `property` and `keyExtractor` are not the same thought .

digorath replied:
I do not know why the warning goes away in the case of you adding a function as the `key` prop of the `FlatList`.
do you maybe get another error or warning instead?

The reason why the warning shows up is not because of the `key` of the `FlatList` but because of its multiple children!

It is not surprising that the warning does not go away, when you do `keyExtractor={ item => item.id}` if the items do not have an `id`, is it? Instead of not defining a `key` for the children, you actively give them. `undefined` instead. ;)

The `key` needs to be a unique value for each of the children. How that value is constructed is up to you. There is no magic answer provided by react/react-native.

Using a `key` which only depends on the `index` can cause issues, as explained in the react documentation, but until you run into those specific issues, it works.

SoleCincis replied:
Not getting other error or warning.  But I have a felling I might have to deal with this when I work on 3rd step: "When reaching the end of the list, fetch more entries of the chosen subject and add them to the list."

digorath commented on components/StyledText.js:
```diff
@@ -2,5 +2,5 @@ import * as React from 'react';
 
 export function MonoText(props) {
-  return <Text {...props} style={[props.style, { fontFamily: 'Inter-Black' }]} />;
+  return <Text {...props} style={[props.style]} />;
```
what is the purpose of `MonoText`?

SoleCincis replied:
at the moment it has no purpose. Before it was use for applying the same fontFamily and style flexibility to all the text components used across the app. Correct ? 
I had to delete the fontFamily cause was creating issues. 
Hopefully I will give a purpose to this component once again when I'm done with the current exercise.

digorath commented on package.json:
```diff
@@ -12,33 +12,33 @@
+        "@react-navigation/native": "^5.10.0",
+        "@react-navigation/stack": "^5.10.0",
+        "@react-navigation/web": "~1.10.0-alpha.9",
+        "expo": "^42.0.0",
```
why is an expo upgrade part of this PR?

SoleCincis replied:
I really do not know how this is part of this PR . expo update is now merged in master and before was on a separate branch which was the base for this branch .  I can see that , the first 3 commits in this PR actually belongs to the expo update branch , reason could be that I have done a reset and not rebase ? 

How I would try to clean it is : doing a cherry-pick of the commits that belongs to  books-API-request branch , into a fresh created branch based on master and then I would need to rebase my other branch for the book list exercise on this last fresh created branch , because they belong to each other. 
Finally I'll be able to delete this books-API-request branch . 

Asking myself if it's all worth it , your opinion on this matter will be decisive

digorath replied:
`reason could be that I have done a reset and not rebase` what did you reset to what?

SoleCincis replied:
I did a soft reset this branch feature/Books-API-request on chore/upgread-SDK-version , which is now merged to master, instead of doing a rebase.

digorath replied:
i don't quite know why the 3 commits were shown in this PR as the commits (even with the same commit hashs) where already in master. usually they do not show up in the diff then.

maybe it was due to resetting, but i don't think so.

it was probably because of you targeting master with this PR, when the sdk pr was not merged yet, and github did not update the diff with the master after that one was merged.

if you had targeted this branch on the sdk branch, the diff would not have included the commits in the beginning, and github would have retargeted this PR to master, once the other branch was merged into master. (!! CAREFUL: GITLAB MIGHT HANDLE THIS DIFFERENTLY !!)

anyways. your suggestion on how to resolve this issue would have been a lot of hassle, and would separate all the discussions/PR from the branch etc.

since they had the same commit hashs, there was no need for drastic measures. it might have been enough to simply change target back and forth to master, but i did a rebase onto master, which updated the diff as well (and added the merge commit of merging sdk updates into master to the branch). nothing else changed though.

digorath replied:
also: please always think, if it is really necessary, when you reset branches. resetting, rebasing and amending commits needs to be done thoughtfully, because it alters the history of changes.
it might make them look better, keep them more readable etc, but it can cause problems if done incorrectly.

digorath commented on screens/BooksScreen.js:
```diff
@@ -0,0 +1,89 @@
+  }, []);
+
+  //looking for ISBN
+  const availabilities =  data.map((book) => {
```
your next step was to handle the isbn inside of the `renderItem` and no longer outside of it. why is it still here?

SoleCincis replied:
thanks for pointing it out . I was stuck on this for way to long... now solved on 02bf2fed85b876743338afb291853980ce0fcc3f

SoleCincis committed:
feat(BooksScreen): isbn for each list item

- like authors isbn is found using item
- and no need to loop to find it thanks to item which is doing the job for me already
 - if item.availability is an object pass it otherwise pass an enpty sting
  -  second conditions to be sure that isbn is a sting otherwise I return nothing

SoleCincis committed:
feat(BooksScreen): isbn for each list item

- like authors isbn is found using item
- and no need to loop to find it thanks to item which is doing the job for me already
 - if item.availability is an object pass it otherwise pass an enpty sting
  -  second conditions to be sure that isbn is a sting otherwise I return nothing

digorath commented:
when trying to install dependencies of your project, i get this. any thoughts?

![Screenshot 2021-11-12 at 11 06 06](https://user-images.githubusercontent.com/59824597/141449334-13becafc-cd4f-48be-aa8d-a570451b0483.png)

digorath commented on screens/BooksScreen.js:
```diff
@@ -0,0 +1,69 @@
+      : `[author: ${item.authors[0].name}]`;
+
+    const isbnArray = typeof item.availability === "object" ? item.availability : '' ;
+    const isbn = typeof isbnArray.isbn === "string" ? `[ISBN: ${isbnArray.isbn}]` : null ;
```
this is a very roundabout way of doing it, but it works.

however it is confusing to read and should be changed.

what i dislike about this solution:
* `item.availability` can be either `undefined` or an `object`.
* if it is an `object`, you keep it, if it is `undefined` you turn it into a `string`.
* then you check if an entry of the `isbnArray` (which is now either a `string` or an `object`, is a `string`
  * it can be easily misread when taking a quick glance, that you check if `isbnArray` itself is a string, as that is one of the cases you specifically introduced in the previous line
  * the likelihood of misreading is further increased by the fact that the `isbn` key does not exist on a `string` and therefore checking the type of it is not necessarily intuitive
* also: what happens if `item.availability` is `null` instead of `undefined`? (currently that does not seem to occur)

SoleCincis replied:
This happens if `item.availability` is `null` instead of `string` 
![IMG_7AF730E53B4C-1](https://user-images.githubusercontent.com/43641321/141811888-172f585f-14f0-48ae-aadb-1d91ed35fa49.jpeg)

I agree with you about the not immediate readability but it's a good compromise with the several lines of code that I would write using an iteration, besides I guess this is a safe way to catch isbn, because I'm eliminating  all the non string types.

Why would you consider this a  "**very** roundabout" ? What makes it **very** ? 

Thanks for your feedback, unfortunately I'm not finding anything in your comment that helps me to find a different way of writing this code, I'd be happy to talk deeper about this in tomorrow' call :)

SoleCincis committed:
Merge branch 'feature/Books-API-request' of github.com:SoleCincis/tracker-App into feature/Books-API-request

SoleCincis commented:
> when trying to install dependencies of your project, i get this. any thoughts?
> 
> ![Screenshot 2021-11-12 at 11 06 06](https://user-images.githubusercontent.com/59824597/141449334-13becafc-cd4f-48be-aa8d-a570451b0483.png)

Seams to be an issue that cams with SDK 40 upgrade . Not sure yet How to fix it ..
